qiskit_optimization/applications/sk_model.py

```diff
@@ -0,0 +1,108 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2021.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+"""An application class for the Sherrington Kirkpatrick (SK) model."""
+
+from typing import Union, List, Optional
+
+import networkx as nx
+import numpy as np
+from docplex.mp.model import Model
+
+from .graph_optimization_application import OptimizationApplication
+from ..algorithms import OptimizationResult
+from ..problems.quadratic_program import QuadraticProgram
+from ..translators import from_docplex_mp
+
+
+class SKModel(OptimizationApplication):
+    r"""Optimization application of the "Sherrington Kirkpatrick (SK) model" [1].
+
+    The SK Hamiltonian over n spins is given as:
+    :math:`H(x)=-1/\sqrt{n} \sum_{i<j} w_{i,j}x_ix_j`,
+    where :math:`x_i\in\{\pm 1\}` and :math:`w_{i,j}\in\{\pm 1\}` are chosen independently and uniformly
+    at random. Notice that there are other variants e.g., with :math:`w_{i,j}` chosen from the normal
+    distribution with mean 0 and variance 1.
+
+    References:
+        [1]: Dmitry Panchenko. "The Sherrington-Kirkpatrick model: an overview",
+        https://arxiv.org/abs/1211.1094
+    """
+
+    def __init__(self, num_of_sites: int, rng: Optional[np.random.Generator] = None):
+        """
+        Constructor for the SK model.
+
+        Args:
+            num_of_sites: number of sites
+            rng: numpy pseudo-random number generator or None.
+            For repeatable experiments use e.g. np.random.default_rng(<SEED>).
+            None results in usage of np.random.default_rng()
+        """
+        if rng is None:
+            self._rng = np.random.default_rng()
+        else:
+            self._rng = rng
+
+        self._num_of_sites = num_of_sites
+        self._graph = nx.complete_graph(self._num_of_sites)
+
+        self.new_disorder()
+
+    def new_disorder(self) -> None:
+        """Generate a new disorder of the SK model."""
+        for _, _, edge_data in self._graph.edges(data=True):
+            edge_data["weight"] = self._rng.choice([-1, 1])
+
+    def to_quadratic_program(self) -> QuadraticProgram:
+        """Convert an SK model problem instance into a
+        :class:`~qiskit_optimization.problems.QuadraticProgram`.
+
+        Returns:
+            The :class:`~qiskit_optimization.problems.QuadraticProgram` created
+            from the SK problem instance.
+        """
+        mdl = Model(name="SK-model")
```

Comment:
Given that the goal seemed to be to generate a bunch of different instances of this class I was wondering with them all having the same name here. On the other hand it may be hard to find a way to make this more unique to each disorder that is simple/easy/short. I guess you can figure things backwards from the objective if you want to get back to the original SK problem given you have a QP that was created from here. Maybe there is no need - but with the one instance changing disorder and emitting a bunch of corresponding QPs was just thinking about a user dealing with that.

Reply:
To be honest I don't know. I would pass that for now. If someone is interested. He/she can add it in another release.